rfc/build-system/build-system.md

```diff
@@ -0,0 +1,698 @@
+
+This document describes:
+
+- changes to the versioning schemes for artifacts and Spinnaker itself
+- a technical description of the build system that will implement these changes
+- how this new build system will interact with the Jenkins-based end-to-end
+  tests
+
+This document does not describe any changes to the end-to-end tests, aside from
+how they interact with the build system. While we would like to eventually move
+these tests into the GitHub Actions CI system and retire Jenkins, that will be a
+separate RFC.
+
+It also does not describe any major changes to the way Deck is built. We will
+continue to build it with Gradle using the same infrastructure as all the other
+microservices. Since Deck is markedly different from the rest of the
+microservices, changing its build system
+[has been a topic of discussion recently](https://docs.google.com/document/d/1ch0wEUNlQwvjAn8ob_HPZ54jYbmmloe2AKbtMRI_JsU/edit?usp=sharing),
+but is not in scope for this document.
+
+It doesn't discuss changes to the Halyard and Spin release process.
+
+## Motivation and Rationale
+
+Currently there are two separate build systems for Spinnaker. These systems
+build the same code, but publish their artifacts to different artifact
+repositories using different version numbers. This is redundant and confusing.
+This document aims to describe a single build system for all community
+artifacts.
+
+One system is orchestrated via GitHub Actions and is used for cross-repository
+dependencies. For example, when `gate` code depends on `fiat` code, the jars for
+that dependency are generated by these builds. This system is also used by the
+plugin system for versioning the plugin APIs.
+
+The other system is used for the artifacts we distribute to the community:
+Docker containers and Debian packages. This system runs in Jenkins and is a
+complicated mash-up of Jenkins and Google Cloud Build configuration, shell
+scripts, Python, and Kotlin. Some of it is checked-in and some of it just lives
+in Jenkins configs or files on the Jenkins server.
+
+These build systems are mostly unaware of each other. They use different version
+numbering schemes, but with identically-shaped version numbers. For example, in
+the past year, they've each generated an echo 2.12.0, but those two 2.12.0s
+aren't related to one another in any way. Their git tags are slightly different
+(`v2.12.0` vs. `version-2.12.0`) and their artifacts are published to different
+Bintray repositories, so there are no conflicts. (Yes, we also have two
+different Bintray repositories.)
+
+## Timeline
+
+This build system will be implemented between the 1.22 release in mid-August and
+the following release (which will no longer be called 1.23) in mid-October.
+
+## Design
+
+### New Versioning Scheme and Release Schedule
+
+Spinnaker releases will switch to date-based versions, so instead of 1.23, we
+will call it 2020.4. The release schedule will change from every 8 weeks to
+every 3 months.
```

Comment:
I like the stability but we'd probably need to relax the definition of what can be backported to the latest release. I don't think it'd be acceptable to wait a quarter to get a fix that is not technically a regression.

Reply:
Agreed, I think we'd probably need to re-think this in the context of less frequent releases.  I think we should still have a reasonably restrictive policy (not allowing features or overly invasive bugs) but probably would need to allow simple, non-invasive bug fixes even if they do not fix regressions.

Reply:
+1. If we only change "must fix a regression" to "must fix a bug," the remaining criteria around exhaustive documentation and test coverage should hopefully be enough to ensure backports are sufficiently simple and non-invasive.